lib/workers/testbot.js

```diff
@@ -0,0 +1,182 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict'
+
+const _ = require('lodash')
+const Bluebird = require('bluebird')
+const Board = require('firmata')
+const drivelist = Bluebird.promisifyAll(require('drivelist'))
+const fs = Bluebird.promisifyAll(require('fs'))
+const mountutils = Bluebird.promisifyAll(require('mountutils'))
+const imageWrite = require('etcher-image-write')
+const visuals = require('resin-cli-visuals')
+
+const WRITE_DAC_REG = 0x00
+const ENABLE_VOUT_SW = 0x03
+const DISABLE_VOUT_SW = 0x04
+const ENABLE_VREG = 0x07
+const ENABLE_FAULTRST = 0x10
+const SD_RESET_ENABLE = 0x12
+const SD_RESET_DISABLE = 0x13
+
+const LED_PIN = 13
+const HW_SERIAL5 = 5
+const SD_MUX_SEL_PIN = 28
+const USB_MUX_SEL_PIN = 29
+
+const getDevInterface = async (devicePath) => {
+  const devIdLink = '/dev/disk/by-id/usb-Generic_Ultra_HS-SD_MMC_000008264001-0:0'
+
+  if (process.platform === 'linux') {
+    return fs.realpathAsync(devIdLink).catch((error) => {
+      throw new Error(`Could not find testbot. Reason: ${error}`)
+    })
+  }
+
+  if (_.isUndefined(devicePath)) {
+    throw new Error('For non-linux operating systems testbot dev interface has to be provided')
+  }
+
+  return devicePath
+}
+
+module.exports = class TestBotWorker {
+  constructor (title, deviceType, options = {}) {
+    this.title = title
+    this.deviceType = deviceType
+    this.options = options
+
+    this.board = new Board(options.devicePath)
+    this.board.serialConfig({
+      portId: HW_SERIAL5,
+      baud: 9600
+    })
+  }
+
+  sendCommand (command, a_ = 0, b_ = 0) {
+    this.board.serialWrite(HW_SERIAL5, [ command, a_, b_ ])
+  }
+
+  async connectSdToTarget (sw) {
+    this.sendCommand(SD_RESET_ENABLE)
+    await Bluebird.delay(10)
+    this.sendCommand(SD_RESET_DISABLE)
+
+    if (sw) {
+      this.board.digitalWrite(LED_PIN, this.board.LOW)
+      this.board.digitalWrite(SD_MUX_SEL_PIN, this.board.LOW)
+    } else {
+      this.board.digitalWrite(LED_PIN, this.board.HIGH)
+      this.board.digitalWrite(SD_MUX_SEL_PIN, this.board.HIGH)
+    }
+
+    return Bluebird.delay(2000)
+  }
+
+  powerTarget (sw) {
+    sw ? this.sendCommand(ENABLE_VOUT_SW) : this.sendCommand(DISABLE_VOUT_SW)
+  }
+
+  async flash (image) {
+    await this.connectSdToTarget(false)
+
+    // For linux, udev will provide us with a nice id for the testbot
+    const destination = await getDevInterface(this.options.devicePath)
+
+    const drives = await drivelist.listAsync()
+    const drive = _.find(drives, {
+      device: destination
+    })
+
+    if (!drive) {
+      throw new Error(`Cannot find ${destination}`)
+    }
+
+    await mountutils.unmountDiskAsync(drive.device)
+    const driveFileDescriptor = await fs.openAsync(drive.raw, 'rs+')
+    const imageSize = await fs.statAsync(image).get('size')
+
+    const emitter = imageWrite.write({
```

Comment:
@zvin Has been making a lot of progress with the Etcher SDK (https://github.com/resin-io-modules/etcher-sdk), so I believe we can start using this module, and abstract away a lot of this code.

Reply:
I see, maybe it is worth having a different PR for this, as there are the bits of code which use the same logic and I think the refactoring should be done in the same PR.